apiserver/facades/client/highavailability/highavailability.go

```diff
@@ -197,13 +201,33 @@ func (api *HighAvailabilityAPI) enableHASingle(ctx context.Context, spec params.
 	if len(addedUnits) > 0 {
 		addUnitArgs := make([]applicationservice.AddUnitArg, len(addedUnits))
 		for i := range addedUnits {
+			// Try and get the placement for this unit. If it doesn't exist,
+			// then the default behaviour is to create a new machine for the
+			// unit.
+			var placement *instance.Placement
+			if i < len(spec.Placement) {
+				var err error
+				placement, err = instance.ParsePlacement(spec.Placement[i])
+				if err != nil {
+					return params.ControllersChanges{}, errors.Annotate(err, "parsing placement")
+				}
+			}
```

Comment:
I know we don't want to improve existing code (outside the domains), but since this is being added now, why not put it directly into one of the service methods in the app domain?